### PR TITLE
build: add automated releases and pre-commit hooks

### LINKED ISSUE #266: Automated release process
Similarly to https://github.com/standard/eslint-config-standard-with-typescript/issues/840.

We could introduce an automated release process where each commit on `master` get evaluated following [Conventional Commits](https://www.conventionalcommits.org/).

We can use [semantic-release](https://www.npmjs.com/package/semantic-release) for that purpose, it also automatically generates the changelog and publish the new version on `npm`.

### PR DESCRIPTION
**Pull request by Maytha8**
**What is the purpose of this pull request? (put an "X" next to item)**

- [ ] Documentation update
- [ ] Bug fix
- [X] New feature
- [ ] Other, please explain:

**What changes did you make? (Give an overview)**

- Install and configure semantic releases.
- Modify the GitHub Workflow to run `semantic-release`.
- Add commit hooks that run `commitlint`, `standard`, and `ts-standard`. 

**Which issue (if any) does this pull request address?**

- Closes #266
- Closes #288 

**Is there anything you'd like reviewers to focus on?**

Please check that all the configuration is what is needed.
Before merging, make sure the GitHub repository secrets `GH_TOKEN` and `NPM_TOKEN` are set to a GitHub Personal Access Token and NPM Access Token respectively.

### CODE REVIEW

File: .commitlintrc.json [hunk 1]
```diff
@@ -1 +1,6 @@
-{ "extends": ["@commitlint/config-conventional"] }
+{
+  "extends": ["@commitlint/config-conventional"],
+  "rules": {
+    "body-max-line-length": [0]
```
Comment (Divlo) [inline .commitlintrc.json:4]: Why this change?

File: package.json [hunk 2]
```diff
@@ -16,6 +16,10 @@
   "author": "Todd Bluhm",
   "license": "MIT",
   "homepage": "https://github.com/standard/ts-standard",
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/Maytha8/ts-standard.git"
```
Comment (Divlo) [inline package.json:21]: Is that a mistake? 😄

File: package.json [hunk 3]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "ts-standard",
   "description": "TypeScript Standard Style based on StandardJS",
-  "version": "12.0.2",
+  "version": "12.0.3",
```
Comment (Divlo) [inline package.json:4]: Now that we automate it, I think we should not increase the version ourselves.
I think we should use `0.0.0-development` instead.

https://github.com/semantic-release/semantic-release/blob/18bd0c490483008230c57ae6a73d07a09dd00d0d/package.json#L4

File: .releaserc.yml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+            section: Refactoring
+          - type: text
+            section: Testing
+  - "@semantic-release/changelog"
+  - "@semantic-release/github"
+  - "@semantic-release/npm"
+  - - "@semantic-release/git"
+  - assets:
+      - CHANGELOG.md
+      - package.json
```
Comment (Divlo) [inline .releaserc.yml:44]: That might be best to not have any assets.
The changelog will be published with the GitHub Release, what do you think? :smile:

File: .github/workflows/nodejs.yml [hunk 5]
```diff
@@ -1,31 +1,49 @@
...
-
-      - name: 'Install Dependencies'
-        run: 'npm install'
-
-      - run: 'npm run lint:commit -- --to "${{ github.sha }}"'
-      - run: 'npm run lint:standard'
-      - run: 'npm run lint:ts-standard'
-      - run: 'npm run test'
+      - run: npm install
+      - run: npm run lint:commit -- --from "7b4219a67b4fe94b1a1487b7d545d0cc7742f770" --to "${{ github.sha }}"
```
Comment (Divlo) [inline .github/workflows/nodejs.yml:39]: Why adding this `from` flag with a hard-coded commit hash?

File: .github/workflows/nodejs.yml [hunk 6]
```diff
@@ -1,31 +1,49 @@
-name: 'tests'
-
-on:
-  push:
-    branches: [master]
-  pull_request:
-    branches: [master]
+on: [push, pull_request]
```
Comment (Divlo) [inline .github/workflows/nodejs.yml:1]: I'm not sure.
Should we separate in another workflow the publish step? We don't want to publish for PRs, only for commits on master. :smile:

**Review by theoludwig — CHANGES_REQUESTED** [review]
Thanks for your contribution, we appreciate it!

I left some comments to open the discussion. :smile:

File: package.json (hunk 2 above)
Comment (Maytha8) [reply package.json]: Yep, I had it set to my fork so I could test commitlint, but it looks like I forgot to change it. This package.json property is needed as part of commitlint's configuration, so it can create releases automatically on the correct GitHub repo.

File: .github/workflows/nodejs.yml (hunk 5 above)
Comment (Maytha8) [reply .github/workflows/nodejs.yml:39]: This is because some old commits in the commit history don't follow the format and cause commitlint to error, as you can see below.
```
$ npm run lint:commit -- --to "79b63eb"

> ts-standard@12.0.3 lint:commit
> commitlint --to 79b63eb

⧗   input: feat(package): Allow multiple projects to be used (#163)
✖   subject must not be sentence-case, start-case, pascal-case, upper-case [subject-case]

✖   found 1 problems, 0 warnings
ⓘ   Get help: https://github.com/conventional-changelog/commitlint/#what-is-commitlint

⧗   input: fix(badges): Fix readme file badges
✖   subject must not be sentence-case, start-case, pascal-case, upper-case [subject-case]

✖   found 1 problems, 0 warnings
ⓘ   Get help: https://github.com/conventional-changelog/commitlint/#what-is-commitlint

⧗   input: fix(typescript): Try to make typescript a peerDependency
✖   subject must not be sentence-case, start-case, pascal-case, upper-case [subject-case]

✖   found 1 problems, 0 warnings
ⓘ   Get help: https://github.com/conventional-changelog/commitlint/#what-is-commitlint

⧗   input: Initial commit
✖   subject may not be empty [subject-empty]
✖   type may not be empty [type-empty]

✖   found 2 problems, 0 warnings
ⓘ   Get help: https://github.com/conventional-changelog/commitlint/#what-is-commitlint
```
I've hardcoded it from the latest commit on standard/ts-standard as that's the last commit before my commits, and this is when I've setup a precommit hook for commitlint.

File: .releaserc.yml (hunk 4 above)
Comment (Maytha8) [reply .releaserc.yml:44]: # `CHANGELOG.md`
This file already exists in this repo, so I might as well setup semantic-release to update it. 😃
Using a changelog file is also a widely accepted and implemented convention, seen in many major projects, including standard/standard.

# `package.json`
@semantic-release/npm updates the package.json version, so might as well commit that.

File: package.json (hunk 3 above)
Comment (Maytha8) [reply package.json:4]: `semantic-release` will bump the version for you, and `@semantic-release/git` will commit and push the version bump to the git repo (This is why the version bump happened, as I was testing semantic-release). The version won't have to be increased manually. :smile: 

You can see this in action in a test repo I made; here's an example commit of what it will do: 
https://github.com/Maytha8/ts-standard-test/commit/b00d5b3fefff87dac6bf0c4f0dccf3c11d116156#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519
Here you can see that it knew what the next version was and bumped it.

The only reason `semantic-release` have their own version set to `0.0.0-development` is because they don't use `@semantic-release/npm` on their own project (it is merely a default dependency).

If you'd like this behavior of leaving the version at `0.0.0-development`, then we could remove package.json from the `assets` property of `@semantic-release/git` in .releaserc.yml.

File: .commitlintrc.json (hunk 1 above)
Comment (Maytha8) [reply .commitlintrc.json:4]: Some old dependabot commits violate the `body-max-line-length` rule, which is set to 100 by `@commitlint/config-conventional`, but this is fixed when hardcoding the range of commits linted (using `--from` arg) (as seen in https://github.com/standard/ts-standard/pull/289#discussion_r1226158590).

This should probably be reverted if the `--from` argument of commitlint is hardcoded.

File: .github/workflows/nodejs.yml (hunk 6 above)
Comment (Maytha8) [reply .github/workflows/nodejs.yml:1]: Probably. :laughing: I'll separate testing and publishing into 2 workflows.

**Comment by Maytha8** [comment]
@theoludwig What are your thoughts on:
- configuring the workflow to publish to a (new) `dev` tag on npm, set the version to `dev-<timestamp>`, and create a pre-release on GitHub; and
- adding a `release` script to package.json that allows you to easily publish to the `latest` tag and create a distribution release.
